utils/stringforwarder/stringforwarder.go

```diff
@@ -7,59 +7,70 @@ import (
 	"sync/atomic"
 )
 
-// stringForwarder will take messages on a receive() method and forward them to
+// StringForwarder will take messages on a receive() method and forward them to
 // callback, but will drop them if callback() has not finished processing the
 // last message. We will track how many messages we have discarded.
-type stringForwarder struct {
-	callback     func(string)
-	buffer       chan string
-	stopch       chan struct{}
-	started      chan struct{}
+type StringForwarder struct {
+	messages     chan<- string
+	done         chan<- struct{}
 	discardCount uint64
 }
 
-func NewStringForwarder(callback func(string)) *stringForwarder {
+// New returns a new StringForwarder that sends messages to the callback,
+// function, dropping messages if the receiver is not ready.
+func New(callback func(string)) *StringForwarder {
 	if callback == nil {
 		// Nothing to forward to, so no need to start the loop(). We'll
 		// just count the discardCount.
-		return &stringForwarder{}
+		// We create the done channel here even though nothing is listening to
+		// it, to make the code in Stop() more clear that it can't be called
+		// more than once.
+		return &StringForwarder{done: make(chan struct{})}
 	}
-	forwarder := &stringForwarder{
-		callback:     callback,
-		buffer:       make(chan string),
-		stopch:       make(chan struct{}),
+
+	messages := make(chan string)
+	done := make(chan struct{})
+	forwarder := &StringForwarder{
+		messages:     messages,
+		done:         done,
 		discardCount: 0,
 	}
 	started := make(chan struct{})
-	go forwarder.loop(started)
+	go loop(started, done, messages, callback)
 	<-started
 	return forwarder
 }
 
-func (f *stringForwarder) Receive(msg string) {
+// Forward makes a non-blocking send of the message to the callback function.
+// If the message is not received, it will increment the count of discarded
+// messages. Forward is safe to call from multiple goroutines at once.
+// Note that if this StringForwarder was created with a nil callback, all
+// messages will be discarded.
+func (f *StringForwarder) Forward(msg string) {
 	select {
-	case f.buffer <- msg:
+	case f.messages <- msg:
 	default:
 		atomic.AddUint64(&f.discardCount, 1)
 	}
 }
 
-func (f *stringForwarder) Stop() uint64 {
-	if f.stopch != nil {
-		close(f.stopch)
-		f.stopch = nil
-	}
+// Stop cleans up the goroutine running behind StringForwarder and returns the
+// count of discarded messages.  It is an error to call Stop more than once.
+func (f *StringForwarder) Stop() uint64 {
+	close(f.done)
 	return atomic.LoadUint64(&f.discardCount)
 }
 
-func (f *stringForwarder) loop(started chan struct{}) {
-	stopch := f.stopch
+// loop pipes messages from the messages channel into the callback function. It
+// closes started to signal that it has begun, and will clean itself up when
+// done is closed.
+func loop(started chan struct{}, done <-chan struct{}, messages <-chan string, callback func(string)) {
```

Comment:
The reason I made loop into a stanadalone function is that it's a goroutine... it doesn't have any state it is storing, and it shouldn't be implied that it is.  There's two separate concerns, a goroutine looping over channel, and a type that can send messages to that channel.  Linking them to the same object is conceptually incorrect, and unnecessary.  Plus it means the goroutine has receive-only halves of done and messages, making it clear that it should not (and cannot) be the one closing them.

That being said I'll put it back to a method, since that seems to be the more usual way to do it here.